### PR TITLE
Compile surface.c as cpp.

### PR DESCRIPTION
**Pull request by ZehMatt**
As the title suggests. Also slightly cleaned up the code.

### CODE REVIEW

File: src/openrct2/paint/tile_element/Surface.cpp [hunk 1]
```diff
@@ -0,0 +1,1922 @@
...
+        const LocationXY16 offset = viewport_surface_paint_data[i][rotation];
+        const LocationXY16 position = 
+        { 
+            (sint16)(base.x + offset.x), 
+            (sint16)(base.y + offset.y),
+        };
+
+        tile_descriptor& descriptor = tileDescriptors[i + 1];
+        descriptor.tile_element = nullptr;
+        if (position.x > 0x2000 || position.y > 0x2000)
```
Comment (marijnvdwerf) [inline src/openrct2/paint/tile_element/Surface.cpp]: @Gymnasiast Does this have something to do with map size limits?

File: src/openrct2/paint/tile_element/Surface.cpp [hunk 2]
```diff
@@ -0,0 +1,1922 @@
...
+    if (gStaffDrawPatrolAreas != 0xFFFF)
+    {
+        const sint32 staffIndex = gStaffDrawPatrolAreas;
+        const bool is_staff_list = staffIndex & 0x8000;
+        const sint16 x = session->MapPosition.x;
+        const sint16 y = session->MapPosition.y;
+        uint8 staffType = staffIndex & 0x7FFF;
+
+        uint32 image_id     = IMAGE_TYPE_REMAP;
+        uint8  patrolColour = 7;
```
Comment (marijnvdwerf) [inline src/openrct2/paint/tile_element/Surface.cpp]: Use `COLOUR_` constants

File: src/openrct2/paint/tile_element/Surface.cpp [hunk 3]
```diff
@@ -0,0 +1,1922 @@
...
+{
+    0, // GRASS
+    0, // SAND (YELLOW)
+    0, // DIRT
+    FLAG_DONT_SMOOTHEN_SELF, // ROCK
+    FLAG_DONT_SMOOTHEN_SELF, // MARTIAN
+    FLAG_DONT_SMOOTHEN_SELF | FLAG_DONT_SMOOTHEN, // GRASS
+    0, // GRASS_CLUMPS
+    0, // ICE
+    FLAG_DONT_SMOOTHEN_SELF | FLAG_DONT_SMOOTHEN, // GRASS
```
Comment (marijnvdwerf) [inline src/openrct2/paint/tile_element/Surface.cpp]: Comments don't seem to match terrain types

**Comment by duncanspumpkin** [comment]
Why has this commit not come up as a rename?

**Comment by janisozaur** [comment]
@duncanspumpkin git does not store renames, it only tracks changes. There is some magic telling `git diff` that if files are similar enough, it should treat them as a rename. You can change the default percentage of likeness from 50 to something else, say 40, with `git diff -M 40`. This is probably just too different in C than in C++, but I agree having a commit with just the rename would be nicer to track changes.

File: src/openrct2/paint/tile_element/Surface.cpp (hunk 1 above)
Comment (ZehMatt) [reply src/openrct2/paint/tile_element/Surface.cpp]: Also this case is already guarded by map_get_surface_element_at -> map_get_first_element_at which would return NULL. The loop will also continue if the element is null. So basically this check is somewhat useless here.

**Comment by duncanspumpkin** [comment]
@janisozaur as far as i can tell there is not radical changes between the two though

**Comment by janisozaur** [comment]
There's lots of reformatting though, this may confuse git

File: src/openrct2/paint/tile_element/Surface.cpp (hunk 1 above)
Comment (Gymnasiast) [reply src/openrct2/paint/tile_element/Surface.cpp]: 0x2000 is 8192 in decimal. Divide this by 32 to get tile coordinates and you get 256.
8192 is indeed the upper limit on sprite coordinates: it means the end of tile 255. (0 = beginning of tile 0, 32 = end of tile 0, beginning of tile 1, etc.)

So, in short: yes, it has.

**Comment by ZehMatt** [comment]
Is there anything left to do from my end?

**Comment by duncanspumpkin** [comment]
Well it's hard to review as it's a 1700 line change with no easy identification of what has been modified.

**Comment by ZehMatt** [comment]
I can probably cleanup the PR to separate rename and editing if thats required?

**Comment by Gymnasiast** [comment]
That's the way to go, yes: one that renames and compiles as C++ (including any necessary changes for the latter), and one that tackles formatting.

**Comment by ZehMatt** [comment]
Well to compile it we require already quite a few changes if I remember right, I'll look into it this week.

**Comment by ZehMatt** [comment]
As suggested I split up the PR. Shall I do the formatting in this PR or another? Theres also the thing about the memmove which I would like to change but thats probably best in its own PR.

**Comment by Gymnasiast** [comment]
What you have now is correct. You don't need to open separate PRs, as long as you have separate commits.

**Comment by ZehMatt** [comment]
So the memmove would be a thing I can include here? I really want to ditch it, I already tried using indices and it works just fine with the benefit of improved debug build performance

**Comment by Gymnasiast** [comment]
Memmove is probably better left for another PR. I was talking about the formatting.

**Comment by ZehMatt** [comment]
Alritey then. I'll probably add another commit for more const qualifying things.

**Comment by ZehMatt** [comment]
I think this should be good for now.

**Comment by ZehMatt** [comment]
If the checks pass then I'll leave at this point, clean enough I think.

**Pull request merged**